File: LeetCode/BinarySearch/java/searchMatrix/searchMatrix.java
package searchMatrix;

import java.util.Arrays;

// You are given an m x n integer matrix matrix with the following two properties:

// Each row is sorted in non-decreasing order.
// The first integer of each row is greater than the last integer of the previous row.
// Given an integer target, return true if target is in matrix or false otherwise.

// You must write a solution in O(log(m * n)) time complexity.

// Example 1:

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 3
// Output: true
// Example 2:

// Input: matrix = [[1,3,5,7],[10,11,16,20],[23,30,34,60]], target = 13
// Output: false

// Constraints:

// m == matrix.length
// n == matrix[i].length
// 1 <= m, n <= 100
// -104 <= matrix[i][j], target <= 104

enum testResult {
    PASS("\u001B[32mPASS\u001B[0m"),
    FAIL("\u001B[31mFAIL\u001B[0m");

    private final String value;

    testResult(String value) {
        this.value = value;
    }

    public String getValue() {
        return this.value;
    }
}

record searchMatrixRecord(int[][] matrix, int target, boolean expected) {
}

class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {

        int rows = matrix.length;
        int cols = matrix[0].length;
        int row = 0;
        int col = cols - 1;

        while (row < rows && col > -1) {
            int cur = matrix[row][col];
            if (cur == target) {
                return true;
            }
            if (target > cur) {
                row++;
            } else {
                col--;
            }
        }

        return false;
    }

    public static void main(String[] args) {
        searchMatrixRecord[] records = new searchMatrixRecord[] {
                new searchMatrixRecord(new int[][] { { 1, 3, 5, 7 }, { 10, 11, 16, 20 }, { 23, 30, 34, 60 } }, 3, true),
                new searchMatrixRecord(new int[][] { { 1, 3, 5, 7 }, { 10, 11, 16, 20 }, { 23, 30, 34, 60 } }, 13,
                        false),
        };

        int i = 1;
        for (searchMatrixRecord record : records) {
            System.out.println("# Test case " + i++);
            Solution.testSolution(record);
            System.out.println("-".repeat(50));
        }
    }

    private static void testSolution(searchMatrixRecord record) {
        System.out.println("input:\tmatrix: " + Arrays.deepToString(record.matrix()));
        System.out.println("\ttarget: " + record.target());
        System.out.println("expected: " + record.expected());

        boolean res = new Solution().searchMatrix(record.matrix(), record.target());

        System.out.println("result: " + res);
        System.out.println(res == record.expected() ? testResult.PASS.getValue() : testResult.FAIL.getValue());
    }
}

File: LeetCode/BinarySearch/python/searchMatrix.py
from dataclasses import dataclass
from enum import Enum
from typing import List

class Result(Enum):
    PASS = "\033[92mPASS\033[00m"
    FAIL = "\033[91mFAIL\033[00m"

@dataclass
class searchMatrixRecord:
    matrix: List[List[int]]
    target: int
    expected: bool

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows: int = len(matrix)
        cols: int = len(matrix[0])
        row: int = 0
        col: int = cols - 1

        while row < rows and col > -1:
            cur: int = matrix[row][col]

            if cur == target:
                return True
            
            if cur < target:
                row += 1
            else:
                col -= 1
        
        return False

    @staticmethod
    def testSolution(record: searchMatrixRecord) -> None:
        print(f"input:\tmatrix: {record.matrix}")
        print(f"\ttarget: {record.target}")
        print(f"expected: {record.expected}")

        res: bool = Solution().searchMatrix(record.matrix, record.target)
        print(f"result: {res}")
        print(Result.PASS.value if res == record.expected else Result.FAIL.value)

if __name__ == "__main__":
    records: List[searchMatrixRecord] = [
        searchMatrixRecord([[1,3,5,7],[10,11,16,20],[23,30,34,60]], 3, True),
        searchMatrixRecord([[1,3,5,7],[10,11,16,20],[23,30,34,60]], 13, False),
    ]

    for i, record in enumerate(records):
        print(f"# Test case {i + 1}")
        Solution.testSolution(record)
        print("-" * 50)